server/server.go

```diff
@@ -52,30 +51,34 @@ func NewServer(cfg *config.Config) (server *Server, err error) {
 
 	log.ConfigureLogger(cfg.LogLevel, cfg.LogFormat, cfg.LogTimestamp)
 
-	dnsServers = append(dnsServers, createUDPServer(getServerAddress(cfg.Port)))
-	dnsServers = append(dnsServers, createTCPServer(getServerAddress(cfg.Port)))
+	type NewServerFunc func(address string) *dns.Server
 
-	if cfg.TLSPort != "" {
-		dnsServers = append(dnsServers, createTLSServer(getServerAddress(cfg.TLSPort), cfg.CertFile, cfg.KeyFile))
+	addServers := func(newServer NewServerFunc, addresses config.ListenConfig) {
+		for _, address := range addresses {
+			dnsServers = append(dnsServers, newServer(getServerAddress(address)))
+		}
 	}
 
-	var httpListener, httpsListener net.Listener
+	addServers(createUDPServer, cfg.DNSPorts)
+	addServers(createTCPServer, cfg.DNSPorts)
 
-	router := createRouter(cfg)
+	addServers(func(address string) *dns.Server {
+		return createTLSServer(address, cfg.CertFile, cfg.KeyFile)
+	}, cfg.TLSPorts)
 
-	if cfg.HTTPPort != "" {
-		if httpListener, err = net.Listen("tcp", getServerAddress(cfg.HTTPPort)); err != nil {
-			return nil, fmt.Errorf("start http listener on %s failed: %w", cfg.HTTPPort, err)
-		}
+	router := createRouter(cfg)
 
-		metrics.Start(router, cfg.Prometheus)
+	httpListeners, err := newListeners("http", cfg.HTTPPorts)
+	if err != nil {
+		return nil, err
 	}
 
-	if cfg.HTTPSPort != "" {
-		if httpsListener, err = net.Listen("tcp", getServerAddress(cfg.HTTPSPort)); err != nil {
-			return nil, fmt.Errorf("start https listener on port %s failed: %w", cfg.HTTPSPort, err)
-		}
+	httpsListeners, err := newListeners("https", cfg.HTTPSPorts)
+	if err != nil {
+		return nil, err
+	}
 
+	if len(httpListeners) != 0 || len(httpsListeners) != 0 {
 		metrics.Start(router, cfg.Prometheus)
```

Comment:
I haven't tested metrics as I don't have the setup for it yet, so please be mindful of this part.

Reply:
I tested it -> works